Implementing Editable Title

emecas opened the pull request:
Fix #4   

   -   Added logic & UI to edit the title and controlling its
       length > 0 and < 30 based on the available area to avoid
       the buttons move or miss-align.

   -   Added a function to persist an object on local storage

   -   Added a new launch/open button using the suggested image icon
       and relocated the logic to launch all the tabs by clicking it.

   -   Moved the 3 icons to the right, added new DIV, added styles
       and updated `setupAccordion ` in `main.js` to support new DIV

Signed-off-by: Emerson Castaneda <emecas@users.noreply.github.com>

emecas committed:
Implement Editable Title

   -   Added logic & UI to edit the title and controlling its
       length > 0 and < 30 based on the available area to avoid
       the buttons move or miss-align.

   -   Added a function to persist an object on local storage

   -   Added a new launch/open button using the suggested image icon
       and relocated the logic to launch all the tabs by clicking it.

   -   Moved the 3 icons to the right, added new DIV, added styles
       and updated setupAccordion in main.js to support new DIV

Signed-off-by: Emerson Castaneda <emecas@users.noreply.github.com>

allen505 commented on popup/warehouse.js:
```diff
@@ -55,27 +59,58 @@ function createList(storeObj) {
 			});
 
+			accButton.addEventListener("keydown" , event => {
+				if (accButton.textContent.length > 30) {
```
From the code written, I understand that it is supposed to not take in any new characters after the 30 char limit. But when I tested it, the new character entered stays, while the old character is removed and the cursor suddenly jumps to the starting point of the text.   
I think a better way to implement would be to take in the string of any length, but before storing, (in the `persist` function) call a function which trims the given string to the cutOff length. A reference implementation can be see [here](https://github.com/allen505/save-for-later/blob/0422715a00253fe4581b6098ecb8709ece688b8e/popup/warehouse.js#L89).  
If you could make the referenced line into a seperate function which can be called in the referenced area and in the `persist` function it would be great.

emecas replied:
The problem is not when i am persisting longer strings, actually the problem is about the visualization during the input, that after 30 characters it starts messing with the buttons, please try by extending the value in the condition and let me know how it behaves for you.

This is what is happening to me when i allow longer strings:

![image](https://user-images.githubusercontent.com/849502/94206959-a11f0100-fe94-11ea-8020-5cd74e0e67ba.png)

I guess it can be solved with styles and an extra DIV as well

emecas replied:
I figured out something else, while testing in Chrome, it does not accept spaces in the title text and the max length before starting moving down the buttons is 19 characters.

Is it same for you? 

In Firefox, it does accept spaces, but the max length is 25 characters (it is not 30)

allen505 commented on popup/warehouse.js:
```diff
@@ -55,27 +59,58 @@ function createList(storeObj) {
+				}
+			});
+
+			accButton.addEventListener("focusout", () => {
```
Here the `focusout` event works as intended on Firefox. However on Chrome things got weird.  

From my testing done on Chrome this is what I think is happening, when a click is done outside the extension, the `focusout` event is not raised. The `focusout` event is raised only when the click is done outside the element listening but inside the Extension popup.  
I see only two options to make it work. If you have any idea of your own, please let me know,
1. Apart from checking the `focusout` event  also check another event with an OR condition. The other event could be `onPointerLeave`. More about it can be found [here](https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onpointerleave)
2. Save the text when an `Enter` or `Carriage Return` is received. For this to be the case though, there needs to be some visual to indicate that the title is not yet saved and when it is saved and possibly a tooltip which says "Hit Enter to confirm edit". This is so that people aren't very tech-savy can also easily use the feature  

Again, I'd like to hear your thoughts on this. Is there any alternate elegant way to know when to save it?

emecas replied:
`pointerleave` seems is working as well, except when you are not using the mouse, even with the OR condition. I will keep working on this, since I want to keep this as simple as possible.

emecas replied:
I  did another round last night, i load the extension in Chrome as well and it seems it is working fine on both browsers whit this block:

```
['pointerleave', 'focusout'].forEach( function(e) {
	accButton.addEventListener(e, () => {
		if(obj.title != accButton.textContent){
			obj.title = accButton.textContent;
			persist(id, obj.title, obj.tabs, obj.tags);
		}
	});
});
```
So, no additional action from the user for saving the text would be required

what do you think?

allen505 reviewed (changes requested):
This is fantastic work @emecas . I have some suggestions which you could look into: 
1. I prefer the variable name `openImg` to `opeImg`
2. I have added some comments about the Title length 
3. The focus out event works on Firefox well, but on Chrome it doesn't work very well. I've added more details in the relevant comment

emecas commented:
> 
> 
> This is fantastic work @emecas . I have some suggestions which you could look into:

Thanks for your review and nice words :)

> 
>     1. I prefer the variable name `openImg` to `opeImg`

Done

> 
>     2. I have added some comments about the Title length
> 
>     3. The focus out event works on Firefox well, but on Chrome it doesn't work very well. I've added more details in the relevant comment

for 2 and 3 i have replied on your comments

emecas committed:
Apply review points 1 & 3

   1 - Updated variable name as openImg - [DONE]
   2 - Title length subject - [WIP]
   3 - Added pointerleave EventListener to persist title - [DONE]

Signed-off-by: Emerson Castaneda <emecas@users.noreply.github.com>

allen505 reviewed (changes requested):
Sorry for the delay. The feature works as expected now except for in one situation.
When the pointer is on the save, and the text is edited, the code which checks for `>=29` and `===0` is not being run and as a result it is currently possible to have a save with zero title length.
I think a neat way to avoid this would be to validate the string length only just before the save within the *EventListener*. If the string length is `>=29` we add trailing `...` and if it is `===0` we restore the previous title value. This validation is also done [here](https://github.com/allen505/save-for-later/blob/0422715a00253fe4581b6098ecb8709ece688b8e/popup/warehouse.js#L89). A new function could be made with a String and cutoffLength as parameters.

emecas commented:
> 
> 
> Sorry for the delay. The feature works as expected now except for in one situation.
> When the pointer is on the save, and the text is edited, the code which checks for `>=29` and `===0` is not being run and as a result it is currently possible to have a save with zero title length.

good catch :) 

> I think a neat way to avoid this would be to validate the string length only just before the save within the _EventListener_. If the string length is `>=29` we add trailing `...` and if it is `===0` we restore the previous title value. This validation is also done [here](https://github.com/allen505/save-for-later/blob/0422715a00253fe4581b6098ecb8709ece688b8e/popup/warehouse.js#L89). A new function could be made with a String and cutoffLength as parameters.

I have updated the logic ,  but now aware of 2 different things:

1. I noted that using `===0` , allows an empty string during the visualization and the buttons move affecting the rows under down

Proposed Solution:  updating the condition  to `===1`
 
2.  I have noted as well that since the font used is no mono-space that limit of 29 does not work for any combination of characters, testing in Firefox, if you try a string composes of only `W`characters the limit is 18,  for an string of `w` the limit is  22, for another letter like `q` up 29 works fine, but for a string of  `.` 59 is the limit. 

Proposed Solution:
Adding one more change on the CSS for `accButton` setting  `max-width: 50px;`, that allows overlapping during the input of long strings , but avoids that the buttons move. Unfortunately it seems this only work for Firefox, for Chrome it has produce a different effect, WIP

allen505 commented:
Hi @emecas , could you please update if you are continuing your work on this issue?

emecas commented:
Hi @allen505 I have not touched the PR in a while, I am still using that last verision locally, I can complete the work if you please let me know what is in pending to consider the work as done.

I remenber it behaves a little different depending of the browser.

Thanks for following this up

allen505 commented:
I have also lost some context on what changes were remaining 😀 . There are some very few situations where empty title can be saved, which isn't much of a concern. There are some things which I just noted now though:

1.  In Google Chrome I'm unable to add spaces to the new title. Not sure why this is happening but its important that we fix this
2. If we can add an EventListener to `contentEditable ` for if it is in focus and change its background slightly grey, it would give an indication as to the title being edited. 
3. Try the following:
    - Create a new Save
    - Select all the title to edit
    - Hit backspace (or delete)
    - Try typing a new title
    - Now the previous title will be added to the right of the cursor
   This isn't a big deal, but it would be awesome if fixed.
If points 1 and 2 are fixed, I think we are good to merge it

emecas commented:
> 
> 
> I have also lost some context on what changes were remaining 😀 . There are some very few situations where empty title can be saved, which isn't much of a concern. There are some things which I just noted now though:
> 

@allen505 , Thanks for recapitulating.  

I had some local progresses as well from Oct 2020 that I never pushed, so I started by reviewing those !!

I am just starting to cover every single point you have mentioned here,  I will be reporting progresses soon.

allen505 commented:
Yes, you are right. Making it a div solves that issue. I will have to do further testing to see if there are any other side effects of that. You can leave it as a `button` for now, and I'll fix it at a later point in time.

emecas commented:
> 
> 
> Yes, you are right. Making it a div solves that issue. I will have to do further testing to see if there are any other side effects of that. You can leave it as a `button` for now, and I'll fix it at a later point in time.

OK, if you are right about leaving it like a `button` then I would focus on points 2 and potentially  3 as well,  in order to complete this feature, thanks!!!

allen505 commented:
Hey @emecas Do let me know if you are working on this. If you do not plan on further development in this PR , I am more than willing to merge it and take up any remaining work myself.

emecas commented:
> 
> 
> Hey @emecas Do let me know if you are working on this. If you do not plan on further development in this PR , I am more than willing to merge it and take up any remaining work myself.

hey @allen505  allow me another round at least for resolving conflicts and verify what is in  pending, I just need to make time for it

allen505 commented:
Sure
Also, I'm planning to add a contributors list in the Changelog. Could you please share some details for the same?

_Name you'd want  to be referred to by: 
Any profile link you want to be added:_

emecas commented:
> _Name you'd want to be referred to by:_  

Emerson Castaneda (emecas)  

> _Any profile link you want to be added:_ 

https://github.com/emecas    or   http://twitter.com/emecas

Thx
